ENH: Add LU decomposition

sinhrks opened the pull request:
It is still under work. I'm going to built ``scipy.linalg.lu`` compat API using blocked gaussian elimination.

```
A = np.array([[7, 3, -1, 2], [3, 8, 1, -4], [-1, 1, 4, -1], [2, -4, -1, 6] ])
A
# array([[ 7,  3, -1,  2],
#        [ 3,  8,  1, -4],
#        [-1,  1,  4, -1],
#        [ 2, -4, -1,  6]])

dA = da.from_array(A, chunks=(2, 2))
dA
# dask.array<from-ar..., shape=(4, 4), dtype=int64, chunksize=(2, 2)>

p, l, u = da.linalg.lu(dA)
l
# dask.array<lu-l-6d..., shape=(4, 4), dtype=int64, chunksize=(2, 2)>

p.compute()
# array([[ 1.,  0.,  0.,  0.],
#        [ 0.,  1.,  0.,  0.],
#        [ 0.,  0.,  1.,  0.],
#        [ 0.,  0.,  0.,  1.]])

l.compute()
# array([[ 1.        ,  0.        ,  0.        ,  0.        ],
#        [ 0.42857143,  1.        ,  0.        ,  0.        ],
#        [-0.14285714,  0.21276596,  1.        ,  0.        ],
#        [ 0.28571429, -0.72340426,  0.08982036,  1.        ]])

u.compute()
# array([[ 7.        ,  3.        , -1.        ,  2.        ],
#        [ 0.        ,  6.71428571,  1.42857143, -4.85714286],
#        [ 0.        ,  0.        ,  3.55319149,  0.31914894],
#        [ 0.        ,  0.        ,  0.        ,  1.88622754]])
```

## scipy results

```
p, l, u = scipy.linalg.lu(A)
p
# array([[ 1.,  0.,  0.,  0.],
#        [ 0.,  1.,  0.,  0.],
#        [ 0.,  0.,  1.,  0.],
#        [ 0.,  0.,  0.,  1.]])

l
# array([[ 1.        ,  0.        ,  0.        ,  0.        ],
#        [ 0.42857143,  1.        ,  0.        ,  0.        ],
#        [-0.14285714,  0.21276596,  1.        ,  0.        ],
#        [ 0.28571429, -0.72340426,  0.08982036,  1.        ]])

u
# array([[ 7.        ,  3.        , -1.        ,  2.        ],
#        [ 0.        ,  6.71428571,  1.42857143, -4.85714286],
#        [ 0.        ,  0.        ,  3.55319149,  0.31914894],
#        [ 0.        ,  0.        ,  0.        ,  1.88622754]])
```

mrocklin commented:
I saw your tweets about this and was hoping that dask.array would get a PR :)

cc @jcrist

mrocklin commented on dask/array/linalg.py:
```diff
@@ -328,3 +328,89 @@ def svd(a, name=None):
+    return res
+
+def _lu_horizontal(arr, l):
+    return np.dot(np.linalg.inv(l), arr)
```
Is `np.linalg.inv` efficient in the triangular case?  I did a quick benchmark and I think that `scipy.linalg.solve_triangular` might be faster here.

mrocklin replied:
```python
In [1]: import numpy as np

In [2]: import scipy.linalg

In [3]: x = np.random.random((2000, 2000))

In [4]: y = np.random.random((2000, 2000))

In [5]: p, l, u = scipy.linalg.lu(x)

In [6]: %time np.linalg.inv(l).dot(y)
CPU times: user 7.05 s, sys: 295 ms, total: 7.35 s
Wall time: 1.86 s
Out[6]: 
array([[  0.47500179,   0.49931231,   0.98423146, ...,   0.32664585,
          0.6498253 ,   0.36445447],
       [  0.20188943,   0.42700747,   0.60685892, ...,   0.0924809 ,
          0.6496587 ,   0.09958131],
       [  0.88167014,   0.66419102,   0.75093161, ...,   0.21974697,
          0.19297681,   0.59164822],
       ..., 
       [  7.13030456,  -0.52782879,   0.53777843, ...,   4.07304968,
          1.4036643 ,   4.20918249],
       [  5.25830631,   1.38309588,  -3.68227833, ...,  11.42730652,
          1.06180677,  -0.80925082],
       [ 11.57086192,   9.53704602,  -3.79445665, ...,   8.64013982,
         -2.69850912,  10.15152519]])

In [7]: %time scipy.linalg.solve_triangular(l, y, lower=True)
CPU times: user 743 ms, sys: 7.88 ms, total: 751 ms
Wall time: 755 ms
Out[7]: 
array([[  0.47500179,   0.49931231,   0.98423146, ...,   0.32664585,
          0.6498253 ,   0.36445447],
       [  0.20188943,   0.42700747,   0.60685892, ...,   0.0924809 ,
          0.6496587 ,   0.09958131],
       [  0.88167014,   0.66419102,   0.75093161, ...,   0.21974697,
          0.19297681,   0.59164822],
       ..., 
       [  7.13030456,  -0.52782879,   0.53777843, ...,   4.07304968,
          1.4036643 ,   4.20918249],
       [  5.25830631,   1.38309588,  -3.68227833, ...,  11.42730652,
          1.06180677,  -0.80925082],
       [ 11.57086192,   9.53704602,  -3.79445665, ...,   8.64013982,
         -2.69850912,  10.15152519]])
```

mrocklin commented:
I'm curious about the permutation matrix part of this computation.  How out-of-core is this algorithm?  How does the size of the input array affect the memory requirements of the algorithm?

sinhrks commented:
I may misunderstand what you're asking, but I think it's difficult to perform permutation (pivotting) over the dask chunks and results in error if permutation is needed.

If we do not permute between blocks, the algorithm perform each blocks from left-top to right-bottom in out-of-core. Assuming 3x3 chunks:

1. compute (0, 0) chunk
2. compute (0, 1), (0, 2), (1, 0), (2, 0) chunks in parallel/out-of-core
3. compute (1, 1) chunk
4. compute (2, 1) (1, 2) chunks in parallel/out-of-core
5. compute (2, 2) chunk

  - http://homepages.math.uic.edu/~jan/mcs572/parallelLU.pdf

My main interest is to use LU as intermediate results to get inverse matrix.

mrocklin commented:
What is the maximum number of chunks that we must have in memory at once?

For example, in the following computation

    x = da.random.random((n, n), chunks=(n/k, n/k))
    P, L, U = da.linalg.lu(x)
    (L.sum() + U.sum()).compute()

What is maximum amount of memory used over time?  At most this number is `n by n` but hopefully it is smaller.  I'll take a look at the LU algorithm paper linked.

sinhrks commented:
Because each computation is performed using blocks of left-side and top-side, maximum number is 2 * (k - 1) + 1 chunks at right-bottom chunk.

I found a bug of current impl when input has 3 or more chunks. Will fix after #886 (as #886 makes current impl little complex).

mrocklin commented on dask/array/linalg.py:
```diff
@@ -328,3 +328,117 @@ def svd(a, name=None):
+            elif i > j:
+                # ToDo: remove _dummy_identity
+                dsk[(name_p, i, j)] = (np.zeros, (a.chunks[0][i], a.chunks[1][j]))
+                dsk[(name_l_internal, i, j)] = (_dummy_identity, (name_lu, i, j))
```
You should be able to replace this with just `(name_lu, i, j)`.  The scheduler accepts direct name aliases.

```python
In [1]: import dask

In [2]: inc = lambda x: x + 1

In [3]: dsk = {'x': (inc, 1), 'y': 'x'}

In [4]: dask.get(dsk, 'y')
Out[4]: 2
```

shoyer replied:
Also, you don't need the parentheses when adding an element with a tuple key to a dict -- Python adds that automatically.

sinhrks replied:
@mrocklin Yes your example should work, and #886 intends to the case like below. Because calculated ``L`` and ``U`` is referred by intermediate calculation.

```
dsk = {('x', 0): np.array([1, 2]),
       ('x', 1): ('x', 0)}
darr = da.Array(dsk, 'x', chunks=(2, ), shape=(4, ))
darr.compute()
# KeyError: ('x', 0)
```


@shoyer Thanks, will fix that.

mrocklin commented:
This looks pretty cool so far.  Image from current state (after removing `_dummy_identity`)

```python
In [1]: import dask.array as da
In [2]: x = da.random.random((16, 16), chunks=(4, 4))
In [3]: p, l, u = da.linalg.lu(x)
In [4]: from dask import visualize
In [6]: visualize(l, u, filename='dask.png')
Out[6]: <IPython.core.display.Image object>
```

![dask](https://cloud.githubusercontent.com/assets/306380/11999723/dddbd484-aa79-11e5-950f-dcef5f41d5fc.png)

The image corroborates your reasoning about out-of-core.  The algorithm looks like it will be easy to compute with little memory.

@jcrist was working on a Cholesky decomposition a while ago.  This might interest him.

shoyer commented on dask/array/linalg.py:
```diff
@@ -328,3 +328,117 @@ def svd(a, name=None):
+
+    dsk.update(a.dask)
+
+    dsk_p = dsk.copy()
```
Is this necessary? @mrocklin any thoughts on if it's dangerous to share a task dict between multiple arrays?

mrocklin replied:
It's probably not necessary, so far all operations within dask are pure.  However, it's also very cheap; we copy large dicts all the time.

```python
In [1]: import dask.array as da
In [2]: x = da.random.random((10000, 10000), chunks=(100, 100))

In [3]: %time y = ((x + 1)**2).sum(axis=0).mean()
CPU times: user 473 ms, sys: 4.31 ms, total: 477 ms
Wall time: 477 ms

In [4]: %time _ = y.dask.copy()
CPU times: user 6.56 ms, sys: 0 ns, total: 6.56 ms
Wall time: 6 ms

In [5]: len(y.dask)
Out[5]: 40605
```

shoyer commented:
It's pretty awesome that this works (mostly) out of core -- I didn't realize that that was possible! I would definitely document the memory requirements in the doc string for this function.

shoyer commented on dask/array/linalg.py:
```diff
@@ -328,3 +329,93 @@ def svd(a, name=None):
+        else:
+            prevs = [(np.dot, (name_l_internal, i, p),
+                              (name_u_internal, p, i)) for p in range(i)]
+            prevs = reduce(lambda x, y: (operator.add, x, y), prevs)
```
@jcrist Is there an easy way to plug into your new tree-reduction code here? Or would that be pointless?

mrocklin replied:
Is there a reason why we're not using `sum` directly here regardless?

mrocklin replied:
e.g. `prevs = (sum, prevs)`

Although, it's nice to break apart large computations like this into multiple keys.  That helps debugging and diagnostic tools to properly classify what's going on.  I wonder if each element in prevs should be a separate key in the graph followed by a final key for `(sum, prev_keys)`.

shoyer replied:
@mrocklin I think the reason not to use `sum` is that it lets you handle the sum in a streaming fashion? Otherwise you'll need to load each of `prevs` into memory together.

mrocklin replied:
At the moment these are all in the same task, so they'll all be loaded in at once.  If we want to do streaming work then each computation will need to be a separate key-value pair in the dictionary.

mrocklin commented:
Is there anything I can do to help with this PR?

sinhrks commented:
Thanks, I still meet the ``KeyError`` and will check the reason.

mrocklin commented:
I think that I see the problem.

mrocklin commented:
Resolved in #903 I think

mrocklin commented:
Tests pass

mrocklin commented on dask/array/linalg.py:
```diff
@@ -328,3 +329,82 @@ def svd(a, name=None):
+        else:
+            prevs = [(np.dot, (name_l, i, p),
+                              (name_u, p, i)) for p in range(i)]
+            prevs = reduce(lambda x, y: (operator.add, x, y), prevs)
```
I think that this should either be `prevs = (sum, prevs)` or it should be many small tasks within the graph.

sinhrks commented:
@mrocklin Thanks, #903 works. I'll fix it in this weekend to:

- [ ] use ``solve_triangular`` and ``np.sum``
- [ ] add more test cases with larger blocks

sinhrks commented:
Sorry to take a long. Once updated to cover more larger matrices/chunks and looks OK. Could you review?

```
np.random.seed(10)
A = np.random.random_integers(0, 10, (100, 100))
A.shape
# (100, 100)

import dask.array as da
dA = da.from_array(A, chunks=(20, 20))
p, l, u = da.linalg.lu(dA)
np.allclose(p.dot(l.dot(u)).compute(), A)
# True

l.visualize()
```

![2016-01-17 13 46 42](https://cloud.githubusercontent.com/assets/1696302/12375994/ccc199ea-bd20-11e5-87bd-047736ad6bf3.png)

mrocklin commented on dask/array/linalg.py:
```diff
@@ -328,3 +329,98 @@ def svd(a, name=None):
+        except scipy.linalg.LinAlgError:
+            # catch LinAlgError
+            # singular matrix: resolution failed at diagonal
+            return np.linalg.inv(x).dot(y)
```
Under what conditions will the `solve_triangular` function fail?  Are there alternatives to computing the full inverse?

If possible it would be nice to avoid defining functions within other functions.  These take significantly longer to serialize when doing distributed computing.

shoyer replied:
Just using `solve` seems like a better fallback

sinhrks replied:
Thank,s ``solve`` should work and could remove ``np.linalg.inv`` completely from the logic.

mrocklin commented on dask/array/linalg.py:
```diff
@@ -328,3 +329,98 @@ def svd(a, name=None):
+    if xdim != ydim:
+        raise ValueError('input must be a square matrix to perform lu decomposition')
+    if not len(set(a.chunks[0] + a.chunks[1])) == 1:
+        raise ValueError('all chunks must be a square matrix to perform lu decomposition')
```
Can we direct the user to the `rechunk` method here?

ahmadia replied:
Creating a set to check if two objects are the same is not a pattern I'm familiar with.  Would it be more obvious here just to see if the objects pass an equality test?

mrocklin replied:
`a.chunks` ideally has the following form: `((n, n, n, n, n), (n, n, n, n, n))` for some integer `n`.  This check looks sane to me given that form.

sinhrks replied:
OK, updated the error mesagge to indicate ``rechunk``.

mrocklin commented:
@ahmadia this PR may interest you.

mrocklin commented:
Profile plots: https://rawgit.com/mrocklin/f73b2967b224b41a62d2/raw/9057bb64edea67f66ecd50a305e614265d03df57/lu.html

```python
In [1]: import dask.array as da

In [2]: x = da.random.random((10000, 1000), chunks=(1000, 1000))
In [3]: y = x.dot(x.T)
In [4]: p, l, u = da.linalg.lu(y)

In [5]: from dask.diagnostics import ResourceProfiler, Profiler, visualize
In [6]: with ResourceProfiler() as rprof, Profiler() as prof:
    l.sum().compute()
   ...:     

In [7]: visualize([rprof, prof], 'lu.html')
Out[7]: <bokeh.models.plots.GridPlot at 0x7f3290076780>
```

ahmadia commented:
Indeed. Thanks for the ping :)
On Sun, Jan 17, 2016 at 1:24 PM Matthew Rocklin <notifications@github.com>
wrote:

> Profile plots:
> https://rawgit.com/mrocklin/f73b2967b224b41a62d2/raw/9057bb64edea67f66ecd50a305e614265d03df57/lu.html
>
> In [1]: import dask.array as da
>
> In [2]: x = da.random.random((10000, 1000), chunks=(1000, 1000))
> In [3]: y = x.dot(x.T)
> In [4]: p, l, u = da.linalg.lu(y)
>
> In [5]: from dask.diagnostics import ResourceProfiler, Profiler, visualize
> In [6]: with ResourceProfiler() as rprof, Profiler() as prof:
>     l.sum().compute()
>    ...:
>
> In [7]: visualize([rprof, prof], 'lu.html')
> Out[7]: <bokeh.models.plots.GridPlot at 0x7f3290076780>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/blaze/dask/pull/885#issuecomment-172360950>.
>

ahmadia commented:
From a parallel linear algebra standpoint, this is probably most interesting as an algorithm for solving dense (input) LU on low-memory compute nodes (not actually that uncommon of a situation) with the out-of-core pieces of the matrix stored somewhere with relatively high access cost compared to memory.  This has implications for Hadoop-style clusters where the matrix to be solved for is not necessarily predistributed across the entire cluster.

It would be worth doing a performance comparison here to see how the algorithm performs against a completely in-memory solution using just `scipy.linalg.lu` and for different approaches to chunking/threading.

On a single-core machine, the approach seems straightforward.  On a multi-core machine, one may tweak both the number of BLAS/LAPACK threads as well as the number of parallel processes.  My intuition is that it will be fastest to use the largest possible blocks that would still fit in memory.  I also would like to see some indication of the performance overhead to solve LU's out of core, so to speak, and where the costs are, if that's possible.

ahmadia commented:
This conference paper provides more details on out-of-core parallel solver frameworks and is relatively recent: http://conferences.computer.org/sc/2012/papers/1000a062.pdf

mrocklin commented:
This looks good to me.  Merging in 24 hours if no comments.

sinhrks commented:
Ah sorry, I've removed it because ``solve(lower=True)`` is meaningless without specifying ``sym_pos`` is ``True`` (am I misunderstood yet?)

- http://docs.scipy.org/doc/scipy-0.16.0/reference/generated/scipy.linalg.solve.html

For efficient calculation, we should use ``solve_triangular`` as initially pointed out. I found the previous ``LinAlgError`` error is caused by a permutation from the triangular to non-triangular. I've fixed the logic and there is no ``LinAlgError`` in test cases.

I've defined ``_solve_triangular_lower`` as module func as it is clear than including ``bool`` in tuples.

Could you take a look once again?

mrocklin commented:
Ah, I incorrectly assumed that `scipy.linalg.solve(..., lower=True)` was `scipy.linalg.solve_triangular(...)`.  What you have here seems correct to me.  Merging.

This is exciting work!

shoyer commented:
Nice work @sinhrks!
On Wed, Jan 20, 2016 at 7:35 AM Matthew Rocklin <notifications@github.com>
wrote:

> Ah, I incorrectly assumed that scipy.linalg.solve(..., lower=True) was
> scipy.linalg.solve_triangular(...). What you have here seems correct to
> me. Merging.
>
> This is exciting work!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/blaze/dask/pull/885#issuecomment-173240741>.
>

poulson commented:
This is very impressive! Are there any (preliminary) GFlop/s tests before and after going out of core? I have always wondered what the performance would be for running a 100 GB LU off of a 1 TB external hard disk.

mrocklin commented:
No numbers yet.  I agree that this would be interesting.  @jcrist any interest in benchmarking this?
